File: Solutions/0101_SymmetricTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool areSymmetric(TreeNode* left, TreeNode* right) {
        if(!left && !right)
            return true;

        if(!left && right || left && !right)
            return false;
        
        if(left->val == right->val) {
            return areSymmetric(left->right, right->left) && areSymmetric(left->left, right->right);
        }
        return false;
    }

    bool isSymmetric(TreeNode* root) {
        return areSymmetric(root->left, root->right);
    }
};

File: Solutions_Easy/0101_SymmetricTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverseFlipped(self, root: Optional[TreeNode]) -> list[TreeNode]:
        if root is None:
            return []

        return (
            root.val,
            self.traverseFlipped(root.right),
            self.traverseFlipped(root.left),
        )

    def traverse(self, root: Optional[TreeNode]) -> list[TreeNode]:
        if root is None:
            return []

        return root.val, self.traverse(root.left), self.traverse(root.right)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root.left is None:
            return root.right is None

        return self.traverse(root.left) == self.traverseFlipped(root.right)
